File: dynamic_programming/java/WordBreak.java
import java.util.*;
public class WordBreak {
    public static boolean wordBreak(String s, List<String> dict) {

        boolean[] dp = new boolean[s.length() + 1];

        dp[0] = true;
        for(int i=1; i <= s.length(); i++){
            for(int j=0; j < i; j++){
                if(dp[j] && dict.contains(s.substring(j, i))){
                    dp[i] = true;
                    break;
                }
            }
        }

        return dp[s.length()];
    }

    public static void main(String[] args) {
        String[] dictionary = { "apple", "pen" };

        List<String> dict = new ArrayList<>();
        for(String s : dictionary){
            dict.add(s);
        }

        if (wordBreak("applepenapple", dict)) {
            System.out.println("Yes");
        } else {
            System.out.println("No");
        }

    }
}

File: dynamic_programming/python/WordBreak.py
# Time: O(n^2)
# Space: O(n)

def word_break(word, dictionary):
    n = len(word)
    dp = [True if not i else False for i in range(n + 1)]
    for i in range(1, n + 1):
        for j in range(i):
            if dp[j] and word[j:i] in dictionary:
                dp[i] = True
                break
    return dp[-1]

if __name__=='__main__':
    print(word_break("applepenapple", ["apple", "pen"]))